common/src/main/java/com/lightstep/tracer/shared/Span.java

```diff
@@ -37,7 +39,14 @@
 
   @Override
   public void finish() {
-    this.finish(System.currentTimeMillis() * 1000);
+    // Note that this.startTimestampRelativeNanos will be -1 if the user
+    // provide an explicit start timestamp in the SpanBuilder.
+    if (this.startTimestampRelativeNanos > 0) {
+      long durationMicros = (System.nanoTime() - this.startTimestampRelativeNanos) / 1000;
+      this.finish(this.record.getOldest_micros() + durationMicros);
+    } else {
+      this.finish(System.currentTimeMillis() * 1000);
```

Comment:
note that a user can still provide microsecond start and finish timestamps, and we honor both/either.